### DISCUSSION TITLE
running tests with app.state.service

### DISCUSSION THREAD

**Discussion by tarnopolskyi** [fastapi/fastapi#8876] [Questions] [answered]
Hello! Problem, guys
i have some services (class UserService for example) - stored in global app's state: app.state.user_service

i'm running tests and it's fails, i have no idea why

```FAILED tests/test_auth.py::test_create_user - AttributeError: 'State' object has no attribute 'user_service'```

test:
``` test_client = TestClient(app)
def test_create_user():
    response = test_client.post(
        '/user/',
        json={
            'username': test_data['data']['username'],
            'email': test_data['data']['email'],
            'password': test_data['plain_password']
        }
    )
    assert 200 == 200
```

**Comment by chbndrhnns** [comment]
Can you post the relevant parts of the production code, as well?
Also, your assert statement will always evaluate to `True` (but that's another topic).

**Comment by tarnopolskyi** [comment]
> Can you post the relevant parts of the production code, as well?
> Also, your assert statement will always evaluate to `True` (but that's another topic).

yes, i know
i've passed 200 == 200 just for CI passing:)

**Comment by tarnopolskyi** [comment]
> Can you post the relevant parts of the production code, as well?
> Also, your assert statement will always evaluate to `True` (but that's another topic).

```@router.post("/", response_model=UserResponse)
async def register_user(
    request: Request, background_tasks: BackgroundTasks, user: User
):
    try:
        current_user = await request.app.state.user_service.create_user(user)
        return current_user
    except UserExistsException:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="User already exists",
        )```

**Comment by chbndrhnns** [comment]
Where are you creating `user_service` and assigning to `app.state.user_service`? 

I think you can try that directly after creating the `app` instance like

```py
app = FastAPI()
app.state.user_service = object()
```

**Comment by tarnopolskyi** [comment]
> Where are you creating `user_service` and assigning to `app.state.user_service`?
> 
> I think you can try that directly after creating the `app` instance like
> 
> ```python
> app = FastAPI()
> app.state.user_service = object()
> ```

what i have now in ```@app.on_event("startup")```
```app.state.user_service = UserService(r, settings.auth)```

all the code works well as expected when i run uvicorn

**Comment by chbndrhnns** [comment, marked as answer]
I believe you are facing the issue described in https://github.com/tiangolo/fastapi/issues/2943.
`asgi-lifespan` helps you to run the lifecycle events during testing.
